language-plutus-core/src/Language/PlutusCore/DeBruijn.hs

```diff
@@ -0,0 +1,266 @@
+{-# LANGUAGE DerivingStrategies    #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+-- | Support for using de Bruijn indices for term and type names.
+module Language.PlutusCore.DeBruijn (
+                                Ix (..)
+                                , DeBruijn (..)
+                                , TyDeBruijn (..)
+                                , FreeVariableError (..)
+                                , deBruijnTy
+                                , deBruijnTerm
+                                , deBruijnProgram
+                                , unDeBruijnTy
+                                , unDeBruijnTerm
+                                , unDeBruijnProgram
+                                ) where
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Pretty
+import           Language.PlutusCore.Quote
+import           Language.PlutusCore.Type
+
+import           Control.Exception
+import           Control.Lens               hiding (Level, index, ix)
+import           Control.Monad.Except
+import           Control.Monad.Reader
+
+import qualified Data.Bimap                 as BM
+import qualified Data.Text                  as T
+import           Data.Typeable
+
+import           Numeric.Natural
+
+import           GHC.Generics
+
+-- | A relative index used for de Bruijn identifiers.
+newtype Ix = Ix Natural
+    deriving stock Generic
+    deriving newtype (Show, Num, Eq, Ord)
+
+-- | A term name as a de Bruijn index.
+data DeBruijn a = DeBruijn { dbnAttribute :: a, dbnString :: T.Text, dbnIndex :: Ix }
+    deriving (Show, Functor, Generic)
+
+-- | A type name as a de Bruijn index.
+newtype TyDeBruijn a = TyDeBruijn (DeBruijn a)
+    deriving (Show, Functor, Generic)
+instance Wrapped (TyDeBruijn a)
+
+instance HasPrettyConfigName config => PrettyBy config (DeBruijn a) where
+    prettyBy config (DeBruijn _ txt (Ix ix))
+        | showsUnique = pretty txt <> "_i" <> pretty ix
+        | otherwise   = pretty txt
+        where PrettyConfigName showsUnique _ = toPrettyConfigName config
+
+deriving newtype instance HasPrettyConfigName config => PrettyBy config (TyDeBruijn a)
+
+class HasIx a where
+    index :: Lens' a Ix
+
+instance HasIx (DeBruijn a) where
+    index = lens g s where
+        g = dbnIndex
+        s n i = n{dbnIndex=i}
+
+instance HasIx (TyDeBruijn a) where
+    index = _Wrapped' . index
+
+-- Converting from normal names to DeBruijn indices, and vice versa
+
+{- Note [Levels and indices]
+The indices that we use for de Bruijn identifiers are *relative*. However,
+when doing conversions it is easier to work with *absolute* levels, since  that
+way we don't have to adjust them all when we go under a binder.
```

Comment:
I misread "easier to work with *absolute* levels" as "easier to return *absolute* levels" as opposed to "easier to have *absolute* levels in state, even though we return *relative* indices". Maybe state explicitly that a term with indices is returned? You describe calculations below from which it follows that indices are returned, but I think explicit words would be helpful here.

Reply:
Yes, I'll make that clearer.